### PR TITLE
Add `this_machine` Python function

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

This PR adds information about machines such as supercomputers. Machines are defined as YAML files in `support/Machines/` and selected in the CMake configuration. This means we can read the machine file in C++ if we want to do resubmissions from within executables at some point. For now I added a Python function that reads the machine file.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: support/Python/Machines.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    Attributes:
+      Name: A short name for the machine. Must match the YAML file name.
+      Description: A description of the machine. Give some basic context and
+        any information that may help people get started using the machine.
+        Provide links to wiki pages, signup pages, etc., for additional
+        information.
+      DefaultProcsPerNode: Default number of worker threads spawned per node.
+        It is often advised to leave one core per node or socket free for
+        communication, so this might be the number of cores or hyperthreads
+        per node minus one.
```
Comment (knelli2) [inline support/Python/Machines.py:47]: Should we have this taken care of somewhere else? By that I mean, the machine file just specifies how many total cores per node there are, and then somewhere else (maybe when creating submit scripts later on) we specify how many comm/worker cores we want.

The reason I say this is because on the larger machines we have access to (Frontera, Anvil, ...) where there are a ton of cores I found for the BBHs that I actually needed multiple comm cores per physical node, effectively splitting the physical node up into several virtual charm nodes. However, the number of virtual nodes we want to create could be problem dependent (maybe splitting the node in half, maybe 3 virtual nodes, etc...), so in that case we wouldn't want that specified here in the machine file.

Thoughts?

File: support/Python/Machines.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        acceptable formats, see: https://slurm.schedmd.com/sbatch.html#OPT_time
+    """
+    yaml_tag = '!Machine'
+    yaml_loader = yaml.SafeLoader
+    # The YAML machine files can have these attributes:
+    Name: str
+    Description: str
+    DefaultProcsPerNode: int = None
+    DefaultQueue: str = None
+    DefaultTimeLimit: str = None
```
Comment (knelli2) [inline support/Python/Machines.py]: Why make these default to `None`? These seem like important things that should have to be specified.

**Review by knelli2 — CHANGES_REQUESTED** [review]

File: support/Python/Machines.py (hunk 1 above)
Comment (nilsvu) [reply support/Python/Machines.py:47]: Sorry, I probably wasn't super clear about the purpose of these values. I added the following to the docstring:

> Many configuration options for job submission are hardcoded in the submit
> script for the machine (such as the total number of available cores per
> node). Here we provide parameters that we want to adjust when scheduling
> jobs, but that have sensible defaults based on our experience running on the
> machine. Parameters are set to 'None' when they have no sensible default for
> the machine.

Does that help? For the PPN we could delete the default and require users to always specify it. However, I think the default is useful because I don't want to remember every time when I schedule a run how many cores the machine has and what's the recommended number of worker threads on this machine. The default is based on experience. It can still be adjusted when scheduling a job if necessary, e.g. to do scaling tests. If the ideal PPN ends up being very problem-dependent then we can think about a way to set it based on the problem. For the majority of cases I don't want to think about this every time I schedule a run though.

File: support/Python/Machines.py (hunk 2 above)
Comment (nilsvu) [reply support/Python/Machines.py]: See added docs above. I think we should only provide defaults where they make sense.

File: support/Python/Machines.py (hunk 1 above)
Comment (knelli2) [reply support/Python/Machines.py:47]: Yeah that makes it clearer. I guess we'll need a utility to edit submit scripts (and update their parameters) once we decide on how we want submit scripts to look.

File: tests/support/Python/Test_Machines.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        yaml.safe_dump(
+            dict(Machine=dict(Name="TestMachine",
+                              Description="Just for testing")),
+            open(self.machinefile_path, 'w'))
+
+    def test_this_machine(self):
+        with self.assertRaises(UnknownMachineError):
+            this_machine('NonexistentMachinefile.yaml')
+        self.assertEqual(
+            this_machine(self.machinefile_path).Name, "TestMachine")
```
Comment (knelli2) [inline tests/support/Python/Test_Machines.py]: Could you add tests for all the parameters you have inside `Machine`? Even the `None` ones, just so we confirm that everything is getting parsed correctly?

File: tests/support/Python/Test_Machines.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+from spectre.Informer import unit_test_build_path
+
+
+class TestMachines(unittest.TestCase):
+    def setUp(self):
+        self.machinefile_path = os.path.join(unit_test_build_path(),
+                                             'TestMachine.yaml')
+        yaml.safe_dump(
+            dict(Machine=dict(Name="TestMachine",
+                              Description="Just for testing")),
```
Comment (knelli2) [inline tests/support/Python/Test_Machines.py]: Maybe also add a test where you set one of the fields that would have been `None`?

File: support/Python/Machines.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        acceptable formats, see: https://slurm.schedmd.com/sbatch.html#OPT_time
+    """
+    yaml_tag = '!Machine'
+    yaml_loader = yaml.SafeLoader
+    # The YAML machine files can have these attributes:
+    Name: str
+    Description: str
+    DefaultProcsPerNode: Optional[int] = None
+    DefaultQueue: Optional[str] = None
+    DefaultTimeLimit: Optional[str] = None
```
Comment (knelli2) [inline support/Python/Machines.py]: [Can't continue the previous thread, so making a new one]

Personally, I think these are all things we should have to specify in the `MACHINE.yaml` file (as defaults that could potentially be overridden). If you want a default-created submit script for a machine (like you mentioned above, not having to input number of worker threads every time you schedule a job), you'll need actual values for these, not `None`.

**Review by knelli2 — CHANGES_REQUESTED** [review]
You can squash any changes

File: support/Python/Machines.py (hunk 5 above)
Comment (nilsvu) [reply support/Python/Machines.py]: I'm not sure I understand you correctly. These values can currently be specified in the machine file, and then they'll be used as default. But they don't **have to** be specified in case there's no obvious default (like maybe the machine has multiple queues and you always want the user to choose one or something).

File: support/Python/Machines.py (hunk 1 above)
Comment (nilsvu) [reply support/Python/Machines.py:47]: Yea I have that written already. My plan is to treat submit scripts as templates with placeholders that can be substituted. They essentially already are templates, only they are edited by hand. So you instead of the current `SPECTRE_BUILD_DIR=/path/to/build/dir` that's in many submit scripts we write `SPECTRE_BUILD_DIR={{ build_dir }}` and the parameter gets substituted. The syntax for placeholders follows Jinja (https://jinja.palletsprojects.com/en/3.1.x/templates/). The CLI is just this:

```sh
$ spectre schedule INPUT_FILE --num_nodes 4  # ...
```

Comment (nilsdeppe) [reply support/Python/Machines.py:47]: Just for SpEC context: SpEC controls things like PPN from a central file (one that holds _all_ supercomputers, so one per system is much nicer). That so far seems to be working quite well. One thing we do need to make sure we can support is running on systems with mixed nodes like Caltech HPC. Ideally we deal with what types of nodes we get at startup, but I don't know CaltechHPC's exact launching config. E.g. if they let peolpe use a subset of a node. We could just restrict ourselves to having the submit script choose whatever queue is emptiest and submit on those nodes :shrug: Nothing to be solved for this PR, just something to think about. I notice multi-queues are mentioned below: I think we want to give good defaults still, ideally we decide dynamically what queue to choose. I think that'll make it easier for new users to just submit jobs and get a reasonable default.

**Review by nilsdeppe — COMMENTED** [review]
I have no additional requests to @knelli2 but just some things to think about for the future directions

File: support/Python/Machines.py (hunk 5 above)
Comment (knelli2) [reply support/Python/Machines.py]: Ah, ok. I *do* think there are some obvious defaults for these for each machine :) (even the ones with multiple queues). But I don't think this is worth holding up this PR. Leave it as is for now.

Comment (nilsvu) [reply support/Python/Machines.py]: You might be right that we'll always have sensible defaults for these. I removed the optional for now and added some tests. If we encounter a machine where we don't have a sensible default, we can make the field optional.

**Review by knelli2 — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
